src/bsp/src/stm32/hal/decawave/platform/stm32f429zi/src/deca_port.c

```diff
@@ -1,9 +1,12 @@
 #include "deca_port.h"
+#include "deca_device_api.h"
 #include "main.h"
 #include "stm32f4xx_hal_conf.h"
 
-static decaNSSConfig_t g_decaNssConfigA = {DW_NSS_A_GPIO_Port, DW_NSS_A_Pin};
-static decaNSSConfig_t g_decaNssConfigB = {DW_NSS_B_GPIO_Port, DW_NSS_B_Pin};
+static decawaveDeviceConfig_t g_decaNssConfigA = {
+    DW_NSS_A_GPIO_Port, DW_NSS_A_Pin, DW_IRQn_A_GPIO_Port, DW_IRQn_A_Pin, 0, NULL, NULL};
+static decawaveDeviceConfig_t g_decaNssConfigB = {
+    DW_NSS_B_GPIO_Port, DW_NSS_B_Pin, DW_IRQn_B_GPIO_Port, DW_IRQn_B_Pin, 1, NULL, NULL};
```

Comment:
The thing is there will be a maximum number of DWs based on the hardware layout. This part of the code will still need to exist to map each possible DW to its pins, we can't just magically make them appear. Also, using an enum vs just indexing the DWs with an int prevents (not as much as if it were a C++ class enum, but still more than an int) trying to access a DW that doesn't exist.

That being said, having all configs seperatly with a switch to return the good one, wasn't the best of idea's. I've change it for an array of configs where the index maps to the enum value of `DW_A`, `DW_B`, etc. The size of the array is scaled by `DWT_NUM_DW_DEV` defined in the DW API which also tells the API the number of DWs it should have in it's internal array.

In the user code, we will end up having some more dynamic way of choosing the devices, but as this is at hardware level, I don't think there's much more I can do. In the end, if we have 6 ports on the HiveBoard, there will be 6 entries in the array and it is in the user code that we will choose which DW to use based on which one responds when we try to communicate with it.

If you have another idea, feel free to share

Reply:
Ok this seems legit to me. So in the end, it will be the programmer's job to scale the number of DW devices according to what he has plugged on the board. That's absolutely fair four our use-case (if our devices were general-purpose with lots of different configs possible, I would have prefered that the config of the right number de peripherals happened automatically based on some data loaded in the EEPROM/flash on bootup, but we are not Trilliant, so this is absolutely fine 😆 ).